File: server/services/business_event_phase_catalog.go
package services

import "sort"

type BusinessEventPhaseCatalogItem struct {
	Code     string `json:"code"`
	Label    string `json:"label"`
	Semantic string `json:"semantic"`
	Order    int    `json:"order"`
}

var businessEventPhaseCatalog = []BusinessEventPhaseCatalogItem{
	{Code: "draft", Label: "草稿", Semantic: "draft", Order: 0},
	{Code: "pending", Label: "待处理", Semantic: "pending", Order: 1},
	{Code: "scheduling", Label: "排产中", Semantic: "pending", Order: 2},
	{Code: "active", Label: "进行中", Semantic: "active", Order: 3},
	{Code: "done", Label: "已完成", Semantic: "done", Order: 4},
	{Code: "cancelled", Label: "已取消", Semantic: "cancelled", Order: 5},
	{Code: "terminal", Label: "终态", Semantic: "terminal", Order: 6},
	{Code: "custom", Label: "自定义", Semantic: "custom", Order: 7},
}

var allowedBusinessStatusPhases = buildAllowedBusinessStatusPhases()

func buildAllowedBusinessStatusPhases() map[string]struct{} {
	allowed := make(map[string]struct{}, len(businessEventPhaseCatalog))
	for _, item := range businessEventPhaseCatalog {
		allowed[item.Code] = struct{}{}
	}
	return allowed
}

func ListBusinessEventPhaseCatalog() []BusinessEventPhaseCatalogItem {
	items := make([]BusinessEventPhaseCatalogItem, len(businessEventPhaseCatalog))
	copy(items, businessEventPhaseCatalog)
	sort.Slice(items, func(i, j int) bool {
		if items[i].Order == items[j].Order {
			return items[i].Code < items[j].Code
		}
		return items[i].Order < items[j].Order
	})
	return items
}

File: src/features/system-mgmt/workflow-core/data/business-event-phase-catalog.ts
import { z } from 'zod'
import { type BusinessStatus } from './business-event-source-types'

export const businessEventPhaseSemanticSchema = z.enum([
  'draft',
  'pending',
  'active',
  'done',
  'cancelled',
  'terminal',
  'custom',
])

export const businessEventPhaseCatalogItemSchema = z.object({
  code: z.string().trim().min(1),
  label: z.string().trim().min(1),
  semantic: businessEventPhaseSemanticSchema,
  order: z.number().int().nonnegative(),
})

export type BusinessEventPhaseSemantic = z.infer<
  typeof businessEventPhaseSemanticSchema
>

export type BusinessEventPhaseCatalogItem = z.infer<
  typeof businessEventPhaseCatalogItemSchema
>

export type BusinessEventPhaseOption = {
  value: string
  label: string
  semantic: BusinessEventPhaseSemantic
  known: boolean
}

export function deserializeBusinessEventPhaseCatalog(
  input: unknown
): BusinessEventPhaseCatalogItem[] {
  return z
    .array(businessEventPhaseCatalogItemSchema)
    .parse(input)
    .slice()
    .sort((a, b) => (a.order === b.order ? a.code.localeCompare(b.code) : a.order - b.order))
}

export function buildBusinessEventPhaseOptions(
  catalog: BusinessEventPhaseCatalogItem[],
  statuses: Pick<BusinessStatus, 'phase'>[] = []
): BusinessEventPhaseOption[] {
  const options = new Map<string, BusinessEventPhaseOption>()

  for (const item of catalog) {
    options.set(item.code, {
      value: item.code,
      label: item.label,
      semantic: item.semantic,
      known: true,
    })
  }

  for (const status of statuses) {
    const phase = status.phase.trim()
    if (!phase || options.has(phase)) {
      continue
    }
    options.set(phase, {
      value: phase,
      label: `${phase}（未注册）`,
      semantic: 'custom',
      known: false,
    })
  }

  return Array.from(options.values())
}
